internal/memoize/memoize.go

```diff
@@ -0,0 +1,201 @@
+package memoize
+
+import (
+	"context"
+	"sync"
+	"time"
+)
+
+type mapValue[V any] struct {
+	v         *V
+	createdAt time.Time
+}
+
+// Map type implements a caching map. The caching map is a map data
+// type where the keys are of type Key and the values are of generic
+// types.
+//
+// The keys in the map have an expiration time. There is a garbage
+// collection process, that periodically cleans the expired keys.
+//
+// There is an upper limit, how many keys the map can store.
+type Map[V any] struct {
+	m             map[string]mapValue[V]
+	stopped       bool
+	cancel        context.CancelFunc
+	lock          sync.RWMutex
+	keyExpiration time.Duration
+	keyLimit      int
+	gcFrequency   time.Duration
+}
+
+// Key is an interface that defines the methods that the Map keys must
+// implement.
+type Key interface {
+	// GetMemoKey returns *string that can be used as the Map
+	// key. The key is ignored, if the GetMemoKey method returns
+	// nil.
+	GetMemoKey() *string
+}
+
+// New function creates a new Map value. The provided ctx can be used
+// to stop the goroutines of the returned Map value.
+func New[V any](ctx context.Context) *Map[V] {
+	ctx, cancel := context.WithCancel(ctx)
+	m := &Map[V]{
+		m:             map[string]mapValue[V]{},
+		stopped:       false,
+		cancel:        cancel,
+		lock:          sync.RWMutex{},
+		keyExpiration: DefaultKeyExpiration,
+		keyLimit:      DefaultKeyLimit,
+		gcFrequency:   DefaultGCFrequency,
+	}
+	m.gc(ctx)
+	return m
+}
+
+// WithKeyExpiration method sets the key expiration value of a Map value.
+func (m *Map[V]) WithKeyExpiration(expiration time.Duration) *Map[V] {
+	m.keyExpiration = expiration
+	return m
+}
+
+// WithGCFrequence method sets the garbage collection frequence of a
+// Map value.
+func (m *Map[V]) WithGCFrequency(frequency time.Duration) *Map[V] {
+	m.gcFrequency = frequency
+	return m
+}
+
+// WithKeyLilmit method sets the key limit of a map value.
```

Comment:
Add comment: what happens when i decrease the limit below the current size? The "cache" could not take any new values until some expired/are invalidated.